cli/src/main/scala/scala/scalanative/cli/ScalaNativeP.scala

```diff
@@ -32,24 +39,89 @@ object ScalaNativeP extends CaseApp[POptions] {
       System.err.println(s"Ignoring non existing path: $path")
     }
 
+    if (options.nirFiles) deserializeFiles(classpath, args.all)
+    else deserializeClassesOrObjects(classpath, args.all)
+  }
+
+  private def deserializeClassesOrObjects(
+      classpath: List[Path],
+      args: Seq[String]
+  ): Unit = {
     Scope { implicit scope =>
       val classLoader =
         ClassLoader.fromDisk {
           Config.empty.withClassPath(classpath)
         }
 
       for {
-        fileName <- args.all
+        className <- args
       } {
-        classLoader.load(Global.Top(fileName)) match {
+        classLoader.load(Global.Top(className)) match {
           case Some(defns) =>
             defns
               .sortBy(_.name.mangle)
               .foreach { d =>
                 println(d.show)
                 println()
               }
-          case None => fail(s"Not found class/object with name `${fileName}`")
+          case None => fail(s"Not found class/object with name `${className}`")
+        }
+      }
+    }
+  }
+
+  private def deserializeFiles(
+      classpath: List[Path],
+      args: Seq[String]
+  ): Unit = {
+
+    case class VirtDirFile(file: Path, virtDir: VirtualDirectory)
+
+    // Paths obtained from VirtualDirectory.files() are decoded as absolute,
+    // but actually are relative to the classpath.
+    // Users are expected to pass relative paths, like in javap and scalajsp.
+    // This is a workaround to standardize the two to allow comparisons.
+    // It is worth noting that it's impossible to pass actual absolute directory here
+    // and lose that data as it should be impossible for a child of a directory to be root.
+    def normalizedVirtDirPathString(path: Path) = {
+      val fileString = path.toString()
+      if (path.isAbsolute()) fileString.substring(1)
+      else fileString
+    }
+
+    Scope { implicit scope =>
+      val allVirtDirFiles =
+        for {
+          path <- classpath
+          virtDir = VirtualDirectory.real(path)
+          file <- virtDir.files
+        } yield VirtDirFile(file, virtDir)
+
+      for {
+        fileName <- args
+      } {
+
+        val foundFileMaybe = allVirtDirFiles.find {
+          case VirtDirFile(file, virtDir) =>
+            val compString = normalizedVirtDirPathString(file)
+            !Files.isDirectory(file) && compString == fileName
+        }
+
+        foundFileMaybe match {
+          case Some(VirtDirFile(file, virtDir)) =>
+            // Used for an internal NIR compatibility assertion message
+            val bufferName = file.getFileName().toString()
+
+            val fileByteBuffer = virtDir.read(file)
+            val defns = deserializeBinary(fileByteBuffer, bufferName)
```

Comment:
Probably we can try to simplify it a bit. Also thanks to `collectFirst` we would not need to iterate over all files in classpath to list their files if eg. all of them are only defined in the first one. Even though for large number of nir files to list it might be suboptimal (however typical you print only 1-2 files)
 
```suggestion
val cp = classpath.map(VirtualDirectory.real(_))
      def virtualDirHasPath(dir: VirtualDirectory, path: Path): Boolean = {
        dir.files.exists { p =>
          // Paths in jars are always using Unix path denotation
          val relativeInJar = p.toString().stripPrefix("/")
          relativeInJar == path.toString()
        }
      }

      for {
        fileName <- args
        path = Paths.get(fileName)
        content <- cp
          .collectFirst {
            case d if virtualDirHasPath(d, path) => d.read(path)
          }
          .orElse(fail(s"Not found file with name `${fileName}`"))
      } {
        val defns = deserializeBinary(content, fileName)
```

Reply:
Unfortunately passing path from Paths.get to VirtualDirectory.read ended up being problematic, most likely because of the different underlying FileSystems. Because of that i've implemented additional filtering through VirtualDirectory.files to obtain the correct Path, but that may have complicated the code